apps/codelab/src/app/admin/feedback/feedback.component.ts

```diff
@@ -1,12 +1,203 @@
 import { Component, OnInit } from '@angular/core';
+import { AngularFireDatabase, AngularFireList } from '@angular/fire/database';
+import { AngularFireAuth } from '@angular/fire/auth';
+import * as firebase from 'firebase/app';
+import { Message } from '@codelab/feedback/src/lib/message';
+import { BehaviorSubject, combineLatest, Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { GithubService } from './github.service';
+
+type Filter = 'all' | 'done' | 'notDone';
+type Grouping = 'nothing' | 'href' | 'name';
+
+function groupBy(feedback: Array<Message>, grouping: Grouping) {
+  const result = feedback.reduce((comment, item) => {
+    const groupName = item[grouping];
+    comment[groupName] = comment[groupName] || [];
+    comment[groupName].push(item);
+    return comment;
+  }, {});
+
+  return Object.keys(result).map(key => ({ key, value: result[key] }));
+}
+
+function normalize(feedback: Array<any>) {
+  return feedback.map(item => ({
+    ...(item.payload && item.payload.val()),
+    key: item.key
+  }));
+}
+
+function group([feedback, grouping]) {
+  if (grouping === 'nothing') {
+    return [
+      {
+        key: 'Messages',
+        value: feedback
+      }
+    ];
+  }
+  if (grouping === 'name' || grouping === 'href') {
+    return groupBy(feedback, grouping);
+  }
+  throw new Error('Unknown grouping: ' + grouping);
+}
+
+function filter([feedback, filterName, [fromDate, toDate]]) {
+  let result;
+  if (filterName === 'all') {
+    result = feedback;
+  }
+
+  if (filterName === 'done') {
+    result = feedback.filter(message => message.isDone);
+  }
+
+  if (filterName === 'notDone') {
+    result = feedback.filter(message => !message.isDone);
+  }
+
+  const fromMs = fromDate ? new Date(fromDate).getTime() : null;
+  const toMs = toDate ? new Date(toDate).getTime() + 86400000 : null; // add 24hrs to include the day of upper bound
+
+  result = result.filter(msg => {
+    const timestampMs = new Date(msg.timestamp).getTime();
+    return (fromMs ? timestampMs >= fromMs : true) && (toMs ? timestampMs <= toMs : true);
+  });
+  return result;
+}
+
 
 @Component({
   selector: 'codelab-feedback',
   templateUrl: './feedback.component.html',
   styleUrls: ['./feedback.component.css']
 })
 export class FeedbackComponent implements OnInit {
-  constructor() {}
+  messages$: Observable<{ key: string; value: Message }[]>;
+  filter$ = new BehaviorSubject<Filter>('notDone');
+  dateFilter$ = new BehaviorSubject<[string, string]>(['', '']);
+  group$ = new BehaviorSubject<Grouping>('nothing');
+  githubAuth;
+  private feedback$: AngularFireList<any[]>;
+  datesForFilter = { dateFrom: '', dateTo: '' };
+
+  constructor(
+    private database: AngularFireDatabase,
+    private afAuth: AngularFireAuth,
+    private ghService: GithubService
+  ) {
+    afAuth.authState.subscribe(authData => {
```

Comment:
This prob should be moved out to some kind of common service?

When this component was created it was the only place which had login, but now when we have login service, this thing should be generalized

Reply:
It doesn't have to be part of this PR though, can come as a follow up, if you do that, just add a TODO